File: src/cowtree/collection.py
"""Collect unreferenced client images without reclaiming live operation inputs."""

from dataclasses import dataclass
from pathlib import Path
import shutil
import uuid

from pydantic import NonNegativeInt, TypeAdapter

from cowtree.durable import sync_directory, write_record
from cowtree.errors import CowtreeError, CowtreeErrorCode
from cowtree.leaves import ForkRecord, Leaves
from cowtree.lifecycle import DropRecord, Lifecycle, ValidationBusyError
from cowtree.metadata import Metadata
from cowtree.metadata_types import Digest, Operation, Record
from cowtree.projection import GitProjection
from cowtree.publication import publish_directory
from cowtree.seals import RetainedNode, SealRecord
from cowtree.views import ViewRecord
from cowtree.workspace import Workspace
from cowtree.workspace_types import Leaf, PublicationRecord


class Maintenance(Record):
    epochs_removed: NonNegativeInt
    receipts_removed: NonNegativeInt
    objects_removed: NonNegativeInt
    temporary_files_removed: NonNegativeInt
    database_bytes: NonNegativeInt
    wal_bytes: NonNegativeInt


class Collection(Record):
    nodes: list[Digest]
    checks: list[int]
    metadata: Maintenance


class QuarantinedNode(Record):
    node: Digest
    commit: str | None


def leaf_roots(leaf: Leaf) -> set[str]:
    roots = {leaf.node}
    if leaf.pending is not None:
        roots.add(leaf.pending.node)
        if leaf.pending.parent_node is not None:
            roots.add(leaf.pending.parent_node)
        if leaf.pending.validation is not None:
            roots.add(leaf.pending.validation.node)
    return roots


@dataclass(frozen=True)
class Collector:
    workspace: Workspace

    def collect(self) -> Collection:
        with self.workspace.session() as metadata:
            checks: list[int] = []
            lifecycle = Lifecycle(self.workspace)
            for leaf in Leaves(self.workspace).records():
                if leaf.check_candidate is None:
                    continue
                try:
                    with lifecycle.guard(leaf=leaf) as descriptors:
                        directory = self.workspace.root / "operations" / f"drop-{uuid.uuid4().hex}"
                        directory.mkdir(mode=0o700)
                        record = DropRecord(leaf=leaf)
                        write_record(path=directory / "drop.json", record=record)
                        lifecycle.remove(
                            metadata=metadata,
                            directory=directory,
                            record=record,
                            descriptors=descriptors,
                        )
                        checks.append(leaf.id)
                except ValidationBusyError:
                    continue
            protected = self.roots()
            removed: list[str] = []
            for directory in sorted((self.workspace.root / "nodes").iterdir()):
                if directory.name in protected:
                    continue
                commit = None
                if (directory / "node.json").exists():
                    commit = self.workspace.nodes.read(identity=directory.name).git_commit
                write_record(
                    path=self.workspace.root / "trash" / f"{directory.name}.json",
                    record=QuarantinedNode(node=directory.name, commit=commit),
                )
                removed.append(directory.name)
            self.recover_locked(metadata=metadata)
            self.artifacts()
            self.workspace.pin_origins(metadata=metadata)
            report = metadata.call(Operation.MAINTAIN).decode(
                "maintenance", TypeAdapter(Maintenance)
            )
            result = Collection(nodes=removed, checks=checks, metadata=report)
        return result

    def recover_locked(self, metadata: Metadata) -> None:
        """Finish interrupted quarantine before a new operation can acquire a node."""
        markers = sorted((self.workspace.root / "trash").glob("*.json"))
        if not markers:
            return
        protected = self.roots()
        projection = GitProjection(
            Leaves(self.workspace).repository(descriptors=metadata.descriptors)
        )
        for marker in markers:
            node = QuarantinedNode.model_validate_json(marker.read_bytes())
            if node.node != marker.stem:
                raise CowtreeError(
                    CowtreeErrorCode.COMMAND_FAILED, "quarantined node identity mismatch"
                )
            source = self.workspace.root / "nodes" / node.node
            directory = self.workspace.root / "trash" / node.node
            if node.node in protected:
                raise CowtreeError(
                    CowtreeErrorCode.COMMAND_FAILED, "quarantined node is still referenced"
                )
            if source.exists():
                publish_directory(source=source, target=directory)
            if node.commit is not None:
                projection.remove_ref(
                    reference=f"refs/cowtree/nodes/{node.node}", expected=node.commit
                )
            if directory.exists():
                shutil.rmtree(directory)
                sync_directory(path=directory.parent)
            marker.unlink()
            sync_directory(path=marker.parent)

    def roots(self) -> set[str]:
        current = Workspace.open(root=self.workspace.root)
        roots = {current.config.initial, current.config.warm_tip}
        for leaf in Leaves(self.workspace).records():
            roots.update(leaf_roots(leaf))
        for path in (self.workspace.root / "retained").glob("*.json"):
            retained = RetainedNode.model_validate_json(path.read_bytes())
            if path.stem != retained.node:
                raise CowtreeError(CowtreeErrorCode.COMMAND_FAILED, "retention identity mismatch")
            roots.add(retained.node)
        for directory in (self.workspace.root / "operations").iterdir():
            if (directory / "fork.json").exists():
                roots.add(
                    ForkRecord.model_validate_json((directory / "fork.json").read_bytes()).node
                )
            elif (directory / "seal.json").exists():
                record = SealRecord.model_validate_json((directory / "seal.json").read_bytes())
                roots.update({record.node, record.before.node})
            elif (directory / "view.json").exists():
                view = ViewRecord.model_validate_json((directory / "view.json").read_bytes())
                roots.update(leaf_roots(view.before))
                if view.after is not None:
                    roots.update(leaf_roots(view.after))
            else:
                raise CowtreeError(
                    CowtreeErrorCode.COMMAND_FAILED, "unrecognized operation blocks collection"
                )
        return roots

    def artifacts(self) -> None:
        receipts = sorted(
            (self.workspace.root / "receipts").glob("*.json"),
            key=lambda path: int(path.stem),
            reverse=True,
        )
        protected_logs: set[Path] = set()
        for path in receipts[:128]:
            record = PublicationRecord.model_validate_json(path.read_bytes())
            protected_logs.add(Path(record.validation.log))
        for path in receipts[128:]:
            path.unlink()
        for leaf in Leaves(self.workspace).records():
            if leaf.check_candidate is not None:
                protected_logs.add(self.workspace.root / "checks" / f"{leaf.id}.log")
            if leaf.pending is not None and leaf.pending.validation is not None:
                protected_logs.add(Path(leaf.pending.validation.log))
        logs = sorted(
            (self.workspace.root / "checks").glob("*.log"),
            key=lambda path: path.stat().st_mtime_ns,
            reverse=True,
        )
        for path in logs[64:]:
            if path not in protected_logs:
                path.unlink()
        live = {leaf.id for leaf in Leaves(self.workspace).records()}
        for path in (self.workspace.root / "checks").glob("*-install"):
            if int(path.name.removesuffix("-install")) not in live:
                shutil.rmtree(path)
        sync_directory(path=self.workspace.root / "receipts")
        sync_directory(path=self.workspace.root / "checks")

File: crates/workspace/src/collection.rs
// Copyright (c) 2026 Windsor Nguyen

//! Quarantine unreachable images while live operation records pin their inputs.

use std::{collections::BTreeSet, fs};

use cowtree_metadata::{LeafId, Maintenance};
use serde::{Deserialize, Serialize};

use crate::{
    CommitId, Error, Leaf, NodeId, Result, Workspace, durability,
    error::Issue,
    leaves::ForkRecord,
    paths, records,
    seals::{RetainedNode, SealRecord},
    session::Session,
    types::PublicationRecord,
    views::ViewRecord,
};

#[derive(Serialize)]
pub struct Collection {
    /// Checkpoint images retired by this pass.
    pub nodes: Vec<NodeId>,
    /// Inactive validation worktrees retired by this pass.
    pub checks: Vec<LeafId>,
    /// Independent authority reclamation receipt.
    pub metadata: Maintenance,
}

#[derive(Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
struct QuarantinedNode {
    /// Checkpoint whose directory has been made ineligible for new readers.
    node: NodeId,
    /// Exact private ref target when a complete node exists.
    commit: Option<CommitId>,
}

impl Workspace {
    pub fn collect(&self) -> Result<Collection> {
        let mut session = self.lock()?;
        let mut checks = Vec::new();
        for leaf in session.leaves()? {
            if leaf.check_candidate.is_none() {
                continue;
            }
            match session.begin_drop(leaf.clone()) {
                Ok(()) => checks.push(leaf.id),
                Err(Error::State { issue: Issue::ValidationBusy, .. }) => {}
                Err(error) => return Err(error),
            }
        }
        let protected = session.roots()?;
        let mut removed = Vec::new();
        for directory in paths::entries(&self.root.join("nodes"))? {
            let name = directory
                .file_name()
                .and_then(|name| name.to_str())
                .ok_or_else(|| Issue::NonUtf8.at(&directory))?;
            let node = NodeId::parse(name)?;
            if protected.contains(&node) {
                continue;
            }
            let commit = if directory.join("node.json").exists() {
                Some(session.nodes().read(&node)?.git_commit)
            } else {
                None
            };
            records::write(
                &self.root.join("trash").join(format!("{}.json", node.as_str())),
                &QuarantinedNode { node: node.clone(), commit },
            )?;
            crate::fault::checkpoint("after-quarantine-intent");
            removed.push(node);
        }
        session.recover_collection()?;
        session.collect_artifacts()?;
        session.pin_origins()?;
        Ok(Collection { nodes: removed, checks, metadata: session.authority.maintain()? })
    }
}

impl Session {
    pub fn recover_collection(&self) -> Result<()> {
        let markers: Vec<_> = paths::entries(&self.workspace.root.join("trash"))?
            .into_iter()
            .filter(|path| path.extension().is_some_and(|extension| extension == "json"))
            .collect();
        if markers.is_empty() {
            return Ok(());
        }
        let protected = self.roots()?;
        for marker in markers {
            let node: QuarantinedNode = records::read(&marker)?;
            if marker.file_stem().and_then(|name| name.to_str()) != Some(node.node.as_str()) {
                return Err(Issue::NodeIdentity.at(&marker));
            }
            if protected.contains(&node.node) {
                return Err(Issue::IncompleteRecord.at(&marker));
            }
            let source = self.workspace.root.join("nodes").join(node.node.as_str());
            let target = self.workspace.root.join("trash").join(node.node.as_str());
            if source.exists() {
                durability::publish_directory(&source, &target)?;
            }
            if let Some(commit) = node.commit {
                self.repository()
                    .remove_ref(&format!("refs/cowtree/nodes/{}", node.node.as_str()), &commit)?;
            }
            if target.exists() {
                records::remove_directory(&target)?;
            }
            fs::remove_file(&marker).map_err(|error| Error::io(&marker, error))?;
            records::sync_directory(&self.workspace.root.join("trash"))?;
        }
        Ok(())
    }

    fn roots(&self) -> Result<BTreeSet<NodeId>> {
        let mut roots = BTreeSet::from([
            self.workspace.config.initial.clone(),
            self.workspace.config.warm_tip.clone(),
        ]);
        for leaf in self.leaves()? {
            leaf_roots(&leaf, &mut roots);
        }
        for path in paths::entries(&self.workspace.root.join("retained"))? {
            if path.extension().is_none_or(|extension| extension != "json") {
                continue;
            }
            let retained: RetainedNode = records::read(&path)?;
            if path.file_stem().and_then(|name| name.to_str()) != Some(retained.node.as_str()) {
                return Err(Issue::NodeIdentity.at(&path));
            }
            roots.insert(retained.node);
        }
        for directory in paths::entries(&self.workspace.root.join("operations"))? {
            if directory.join("fork.json").exists() {
                let record: ForkRecord = records::read(&directory.join("fork.json"))?;
                roots.insert(record.node);
            } else if directory.join("seal.json").exists() {
                let record: SealRecord = records::read(&directory.join("seal.json"))?;
                roots.extend([record.node, record.before.node]);
            } else if directory.join("view.json").exists() {
                let record: ViewRecord = records::read(&directory.join("view.json"))?;
                leaf_roots(&record.before, &mut roots);
                if let Some(after) = record.after {
                    leaf_roots(&after, &mut roots);
                }
            } else {
                return Err(Issue::UnknownOperation.at(&directory));
            }
        }
        Ok(roots)
    }

    fn collect_artifacts(&self) -> Result<()> {
        let root = &self.workspace.root;
        let mut receipts = Vec::new();
        for path in paths::entries(&root.join("receipts"))? {
            if path.extension().is_none_or(|extension| extension != "json") {
                continue;
            }
            let version = path
                .file_stem()
                .and_then(|name| name.to_str())
                .and_then(|name| name.parse::<u64>().ok())
                .ok_or_else(|| Issue::IncompleteRecord.at(&path))?;
            receipts.push((version, path));
        }
        receipts.sort_by_key(|(version, _)| std::cmp::Reverse(*version));
        let mut protected = BTreeSet::new();
        for (_, path) in receipts.iter().take(128) {
            let record: PublicationRecord = records::read(path)?;
            protected.insert(record.validation.log);
        }
        for (_, path) in receipts.iter().skip(128) {
            fs::remove_file(path).map_err(|error| Error::io(path, error))?;
        }
        let leaves = self.leaves()?;
        for leaf in &leaves {
            if leaf.check_candidate.is_some() {
                protected.insert(root.join("checks").join(format!("{}.log", leaf.id.get())));
            }
            if let Some(validation) =
                leaf.pending.as_ref().and_then(|pending| pending.validation.as_ref())
            {
                protected.insert(validation.log.clone());
            }
        }
        let mut logs = Vec::new();
        for path in paths::entries(&root.join("checks"))? {
            if path.extension().is_none_or(|extension| extension != "log") {
                continue;
            }
            let modified = fs::metadata(&path)
                .and_then(|metadata| metadata.modified())
                .map_err(|error| Error::io(&path, error))?;
            logs.push((modified, path));
        }
        logs.sort_by_key(|(modified, _)| std::cmp::Reverse(*modified));
        for (_, path) in logs.iter().skip(64) {
            if !protected.contains(path) {
                fs::remove_file(path).map_err(|error| Error::io(path, error))?;
            }
        }
        let live: BTreeSet<_> = leaves.iter().map(|leaf| leaf.id.get()).collect();
        for path in paths::entries(&root.join("checks"))? {
            let Some(name) = path
                .file_name()
                .and_then(|name| name.to_str())
                .and_then(|name| name.strip_suffix("-install"))
            else {
                continue;
            };
            let id = name.parse::<u64>().map_err(|_| Issue::IncompleteRecord.at(&path))?;
            if !live.contains(&id) {
                records::remove_directory(&path)?;
            }
        }
        records::sync_directory(&root.join("receipts"))?;
        records::sync_directory(&root.join("checks"))
    }
}

fn leaf_roots(leaf: &Leaf, roots: &mut BTreeSet<NodeId>) {
    roots.insert(leaf.node.clone());
    if let Some(pending) = &leaf.pending {
        roots.insert(pending.node.clone());
        roots.extend(pending.parent_node.iter().cloned());
        if let Some(validation) = &pending.validation {
            roots.insert(validation.node.clone());
        }
    }
}
